Description: Compatibility patch for Ubuntu and Debian 12
Forwarded: not-needed
Author: Eric Long <i@hack3r.moe>

diff --git a/debian/control b/debian/control
index af47c7b..a5e5ce6 100644
--- a/debian/control
+++ b/debian/control
@@ -15,6 +15,7 @@ Build-Depends:
  libbpf-dev,
  libffi-dev,
  libxdp-dev,
+ linux-source,
 Standards-Version: 4.7.2
 Homepage: https://github.com/hack3ric/mimic
 Rules-Requires-Root: no
@@ -41,12 +42,21 @@ Description: eBPF UDP -> TCP obfuscator
 
 Package: mimic-dkms
 Section: kernel
-Architecture: all
+Architecture: amd64 arm64 powerpc ppc64 ppc64el riscv64
 Provides: mimic-modules (= ${binary:Version})
 Conflicts: mimic-modules
 Depends:
  ${misc:Depends},
  ${shlibs:Depends},
+ pahole,
+ bubblewrap,
+ xz-utils,
+ lz4,
+Recommends:
+ bzip2,
+ lzma,
+ lzop,
+ zstd,
 Description: eBPF UDP -> TCP obfuscator (DKMS kernel module)
  Mimic is an experimental UDP to TCP obfuscator designed to bypass UDP
  QoS and port blocking. Based on eBPF, it directly mangles data inside
diff --git a/debian/mimic-dkms.install b/debian/mimic-dkms.install
index 552dd47..bec5150 100644
--- a/debian/mimic-dkms.install
+++ b/debian/mimic-dkms.install
@@ -1 +1,7 @@
 kmod/* usr/src/mimic-${env:DEB_VERSION_UPSTREAM}
+
+out/extract-btf usr/lib/mimic
+tools/extract-vmlinux usr/lib/mimic
+tools/vmlinux-to-btf usr/lib/mimic
+
+linux-source-*/tools/bpf/resolve_btfids/resolve_btfids usr/lib/mimic
diff --git a/debian/rules b/debian/rules
index 3f1b343..e4f9b9b 100755
--- a/debian/rules
+++ b/debian/rules
@@ -6,6 +6,13 @@ export DH_VERBOSE = 1
 export DEB_BUILD_MAINT_OPTIONS = hardening=+all
 export DEB_VERSION_UPSTREAM
 
+IS_UBUNTU := $(shell dpkg-vendor --derives-from ubuntu > /dev/null 2>&1; echo $$?)
+
+njobs = 1
+ifneq (,$(filter parallel=%,$(DEB_BUILD_OPTIONS)))
+njobs = $(patsubst parallel=%,%,$(filter parallel=%,$(DEB_BUILD_OPTIONS)))
+endif
+
 %:
 	dh $@
 
@@ -21,10 +28,25 @@ override_dh_auto_configure:
 	cp install/mimic.sysusers debian/mimic.sysusers
 
 override_dh_auto_build:
-	dh_auto_build -- build-cli generate-manpage \
+ifeq ($(IS_UBUNTU),0)
+#	Ubuntu's bpftool relies on current kernel, which is both inconsistent
+#	with Debian and sbuild-unfriendly. Build from kernel source instead.
+	tar xf `bash -c '_tarballs=(/usr/src/linux-source-*.tar.*); echo $${_tarballs[0]}'`
+	make -j$(njobs) -C linux-source-*/tools/bpf/bpftool
+	dh_auto_build -- build-cli build-tools generate-manpage \
+		MODE= \
+		BPF_CC=clang \
+		USE_LIBXDP=1 \
+		BPFTOOL=linux-source-*/tools/bpf/bpftool/bpftool
+else
+	tar xf `bash -c '_tarballs=(/usr/src/linux-source-*.tar.*); echo $${_tarballs[0]}'` \
+		--wildcards 'linux-source-*/tools/' 'linux-source-*/scripts/'
+	dh_auto_build -- build-cli build-tools generate-manpage \
 		MODE= \
 		BPF_CC=clang \
 		USE_LIBXDP=1
+endif
+	make -j$(njobs) -C linux-source-*/tools/bpf/resolve_btfids
 
 override_dh_auto_test:
 
@@ -33,4 +55,5 @@ override_dh_dkms:
 
 override_dh_auto_clean:
 	dh_auto_clean
+	rm -rf linux-source-*
 	rm -f debian/mimic@.service debian/mimic.sysusers
