--- Makefile.orig	2024-02-27 22:28:52.275392806 +0800
+++ Makefile	2024-02-27 22:29:19.936865864 +0800
@@ -48,7 +48,7 @@
 
 all: $(BINARY)
 
-prepare: $(BPFOBJ) $(SUBCMDOBJ)
+prepare: $(SUBCMDOBJ)
 
 $(OUTPUT) $(OUTPUT)/libsubcmd $(LIBBPF_OUT):
 	$(call msg,MKDIR,,$@)
@@ -74,7 +74,7 @@
           -I$(SUBCMD_INCLUDE) \
           $(LIBELF_FLAGS)
 
-LIBS = $(LIBELF_LIBS) -lz
+LIBS = $(LIBELF_LIBS) -lz -lbpf
 
 export srctree OUTPUT HOSTCFLAGS_resolve_btfids Q HOSTCC HOSTLD HOSTAR
 include $(srctree)/tools/build/Makefile.include
@@ -82,9 +82,9 @@
 $(BINARY_IN): fixdep FORCE prepare | $(OUTPUT)
 	$(Q)$(MAKE) $(build)=resolve_btfids
 
-$(BINARY): $(BPFOBJ) $(SUBCMDOBJ) $(BINARY_IN)
+$(BINARY): $(SUBCMDOBJ) $(BINARY_IN)
 	$(call msg,LINK,$@)
-	$(Q)$(HOSTCC) $(BINARY_IN) $(KBUILD_HOSTLDFLAGS) -o $@ $(BPFOBJ) $(SUBCMDOBJ) $(LIBS)
+	$(Q)$(HOSTCC) $(BINARY_IN) $(KBUILD_HOSTLDFLAGS) -o $@ $(SUBCMDOBJ) $(LIBS)
 
 clean_objects := $(wildcard $(OUTPUT)/*.o                \
                             $(OUTPUT)/.*.o.cmd           \
